compiler/optimizing/induction_var_simplification.cc

```diff
@@ -0,0 +1,318 @@
+      break;
+      case HInstruction::kMul:
+       new_val = Int64FromConstant(cst1) * Int64FromConstant(cst2);
+       break;
+      default :
+       UNREACHABLE();
+     }
+      new_cst = graph->GetConstant(type, new_val);
+      return new_cst;
+    } else if (cst1->IsDoubleConstant()) {
+      double new_val =0.0;
+      switch (kind){
+        case HInstruction::kAdd:
+        new_val = cst1->AsDoubleConstant()->GetValue() + cst2->AsDoubleConstant()->GetValue();
+        break;
+        case HInstruction::kSub:
+        new_val = cst1->AsDoubleConstant()->GetValue() - (cst2->AsDoubleConstant()->GetValue());
+        break;
+        case HInstruction::kMul:
+        new_val = cst1->AsDoubleConstant()->GetValue() * cst2->AsDoubleConstant()->GetValue();
+        break;
+        default :
+          UNREACHABLE();
+      }
+      new_cst = graph->GetDoubleConstant(new_val);
+      return new_cst;
+    } else if (cst1->IsFloatConstant()) {
+      float new_val = 0.0;
+      switch (kind) {
+       case HInstruction::kAdd:
+        new_val = cst1->AsFloatConstant()->GetValue() + cst2->AsFloatConstant()->GetValue();
+        break;
+       case HInstruction::kSub:
+        new_val = cst1->AsFloatConstant()->GetValue() - cst2->AsFloatConstant()->GetValue();
+        break;
+       case HInstruction::kMul:
+        new_val = cst1->AsFloatConstant()->GetValue() * cst2->AsFloatConstant()->GetValue();
+        break;
+        default:
+          UNREACHABLE();
+      }
+      new_cst = graph->GetFloatConstant( new_val);
+      return new_cst;
+    }
+    return nullptr;
+}
+static HConstant* GetZeroConstant(HInstruction* input){
+   HConstant * zero = nullptr;
+   DataType::Type type = input->GetType();
+   HGraph* graph = input->GetBlock()->GetGraph();
+   if (DataType::IsIntegralType(type)) {
+      zero = graph->GetConstant(type, 0);
+   } else if (type == DataType::Type::kFloat32) {
+     zero = graph->GetFloatConstant(0.0f);
+   } else if (type == DataType::Type::kFloat64) {
+     zero = graph->GetDoubleConstant(0.0f);
+   }
+   return zero;
+}
+
+static HConstant* GetOneConstant(HInstruction* input){
+   HConstant * one = nullptr;
+   DataType::Type type = input->GetType();
+   HGraph* graph = input->GetBlock()->GetGraph();
+   if (DataType::IsIntegralType(type)) {
+     one = graph->GetConstant(type, 1);
+   } else if (type == DataType::Type::kFloat32) {
+     one = graph->GetFloatConstant(1.0f);
+   } else if (type == DataType::Type::kFloat64) {
+     one = graph->GetDoubleConstant(1.0f);
+   }
+   return one;
+}
+
+HInstruction* GetOutOfLoopInput(HLoopInformation* loop, HPhi* phi) {
+ HInputsRef inputs = phi->GetInputs();
+ for (size_t i = 0; i < inputs.size(); ++i) {
+   HInstruction* input = inputs[i];
+   if (input->GetBlock()->GetLoopInformation() != loop) {
+    return input;
+   }
+ }
+ return nullptr;
+}
+
+bool InductionVarSimplification::IsCandidatePhi(HLoopInformation* loop, HPhi* phi) {
```

Comment:
Put a comment in the corresponding .h.
I think this is the function by which you are identifying the candidate Derived IVs ?